### ISSUE TITLE
Community Contributions

### ISSUE THREAD

**Issue by bwhm** [Joystream/joystream#3473]
Starting with Olympia, we are trying to get the community more involved in the development (and maintainance) of our code base in general, and user facing applications especially.
I think we should ignore mostly the runtime at this stage, and focus mainly on Pioneer especially, as that is both where there is the most low hanging fruit and the product that arguably needs the most new features, but Atlas, and the cli-s (Joystream, Colossus and Argus) should also be good candidates.

We will want to make some of these in to Bounties for the community, and others as tasks for the Builders Working Group. The Bounties would in most/all cases be rather fast and easy, and not require much research or understanding of the funcionality. For the builders, my hope is that although we'll likely benefit from starting small, we can gradually assign them more difficult issues as well.

I'll be happy to debate the specifics, but this is how I've outlined it in my head:
1. Go through all open issues in both repos, and identify issues that fit most/many of the descriptions below:
  - Not urgent - meaning if it isn't done in 2-3 weeks, there is still time to re-assign to someone in the Jsgenesis Team
  - Design is (or will be) ready
  - Not "too" complicated - meaning it wouldn't take more than a days work for someone familiar with the code base
  - Well described or "obvious" SoW - meaning reading
  - Doesn't impact too much else
  - Presumably fairly easy to review
  - etc.
2. For each of these, we assign:
  - a label, I'm thinking `community` or something, that makes it easy to filter.
  - a way to set to set a difficulty level and time estimate.
    - difficulty level:
      - we can choose to make this precise, but I think that would come at the cost of complexity for ourselves and many users. Unless you strongly disagree, I'd say we use a single parameter that bakes in: familiarity with the codebase, understanding of the application/platform, technical skills in general, complexity, and so forth. We assign it a number of 1-5 based on all of these.
    - time estimate:
      - from experience, we know making our own time estimates are hard enough, so trying to set some "generalized" hourly estimate would presumable be even tougher. However, as the opportunity cost here literally is doing it ourselves, I think it's ok to use yourself as the benchmark.
  - some sort of deadline. This should be derived from, but not equal, the point at which we need it done. Keep in mind that it's unlikely that the (first?) PR will be ok to merge and push to production. I'd suggest start from the latest you could re-assign it to someone on the team, and add a week or two at least.
3. We also need to add **some** specific/background information and requirements. This will likely vary a lot from issue to issue, but we need to remember the community are not part of our stand ups or chat. Examples:
  - which branch to push to
  - how they can test their work (we should consider maintaining, or helping to deploy, a staging network for the community)
  - a pointer on where to look for similar work
  - etc.
4. Finally, and this is perhaps the most important part, we need a set of general guidelines and workflow requirements. Although this won't be free, we can't afford to let this be a big cost to the project in terms of manhours spent per output. We need to be strict on this, and not allow endless rounds of reviews that takes longer to address than just doing it yourself. On the other hand, we need to get back to them quickly (this is something we, and me in particular, have struggled to manage) when a response from us required - regardless of whether it's a PR or a follow up questions. We should try to agree on a specific workflow, that will be true across the different repos and products. This will likely need some iterations as the work comes in, but some ground rules, such as having the Lead of the group assign someone to review before anyone in Jsgenesis does, seems obvious.

With pioneer soon to be released with Olympia, we have to go through the many open issues there that wasn't critical for the release anyway. That seems like a good time to create some candidate issues.

I will start racking up issues for the joystream, storage and distributor cli.

cc @kdembler @thesan @Lezek123.

**Comment by singulart** [comment]
The pioneer 2 design is not responsive, meaning, one can't use it on Mobile. If Builders come up with an idea (plan) on how this can be addressed, would that fit in your vision of Community Contribution, that is, would it have a chance to be funded by JSG?

**Comment by bwhm** [comment]
Here is my draft of the respective workflows for community development, for review by the stakeholders.

## Jsgenesis - Creation
1. Identify open issues (or create them) that are suitable tasks for the community. Give it the label `community-dev`.
2. Assign them a value, that represents the expected workload in hours, under the pretext it's a assigned to an experienced dev that understands the project.
3. If the number of hours is >3, add them to the "Community Backlog" column. Else, add it to the "Bounty Backlog" column.

This stops the first part of the flow for Jsgenesis.

## Community
1. The Lead, or some other designated Worker that has been given `triage` access, can now assign the issue to themselves or another Worker. Unless the issues are _very_ closely related, only one issue can be assigned a single Worker at the time. 
2. Once the assigned has started working, move the issue to `Dev in progress`. 
3. Once the issue is completed as per the assigned, they open up a **draft** PR to the `olympia` branch (uno in the issue),  adds a comment on the issue with a link to the PR,  and notifies the Lead/a Worker with `triage` access.
4. The Lead/a Worker with `triage` moves the issue to the `Internal review` column,  and an internal (meaning from the WG) is assigned to review.
5. Once the internal review is approved, the issue is moved to `Code Review`. Jsgenesis reviews within 2 business days.

This should be quite strict, and shouldn't just be testing whether it works or not, but also cover coding conventions and structure (formal and informal) are respected. 

6. Only two rounds of reviews are guaranteed by Jsgenesis. Whereas the first will provide some helpers and tips, the second will only be approve/reject.

## Jsgenesis - Review
1. In the first review, we are in general quite helpful, and the reviewer should be a little more pedagogical than if the PR was made by a member of the Jsgenesis team. However, we remain as strict as we would be wrt quality of work.
2. If the review is accepted as is (regardless of first or second), do a re-evaluation of the workload in terms of hours. Again, from your own (or another member of the Jsgenesis team) point of view.
3. If the second review still requires some improvements, you may choose between:
   i. reject the work outright and close the PR. Should be the case if the work required to fix the issue is _even close_ to the same as making a new PR from scratch, or if code itself is simply not up to our standard.
   ii. pull in the PR, make the (small) changes yourself and push to the workers branch
   iii. provide clear instructions, and if this is not adhered, the PR is closed.


FYI for the community:
If the PR is closed, this will be considered a "Catastrophic error". If the PR is approved, but after 3.i. or 3.ii. the points will get reduced by a factor of 0.5.

**Comment by dmtrjsg** [comment]
### Part 1

> Identify open issues (or create them) that are suitable tasks for the community. Give it the label community-dev.
Yes, label is the way to go for cross-sync on diff boards. Plus there's a Project field that we'd probably populate as well for the issues to appear in backlog automatically! 
> Assign them a value, that represents the expected workload in hours, under the pretext it's a assigned to an experienced dev that understands the project.
- We can try but estimated time vs actual will be rather different.  We are already using story points in our core team; perhaps we could on our end add story points (= complexity); while community devs and testers can add _actual_ time spent? The reward then can be adjusted for time spent in less trivial cases..
> If the number of hours is >3, add them to the "Community Backlog" column. Else, add it to the "Bounty Backlog" column.
- Like the idea behind splitting to two lanes.  Ok we can try with 6hrs and see where we end up with!

Our Pioneer Grooming is set to Thursday 7th, that's when it would be a good time to allocate hourly estimates to all backlog. I'd need to pre-groom it first (add more description and pre-select less tricky ones). There's a list of 100 tasks+ so that's a solid afternoon there on its own!

### Part 2

Overall would need @thesan  to have a look and comment on the details. From my humble perspective that's how I imagine it.  2 days is reasonable..


### Part 3

>push to the workers branch

So we'd need to have a separate branch for community.. I guess Theo would need to think how to handle the wf on the branches and merges lvl.

**Comment by singulart** [comment]
As the Lead of community builders squad, I don't see how the proposed scheme can be put in effect without prior settling of the working conditions and wages. With the current 6,000,000 tJOY / week budget ($120 in equivalent), we can't do much. 

Let me re-iterate: most of the current Builders are experienced engineers, with in-depth Joystream-specific knowledge, and I can't stress enough how important to keep them in project. As a Squad, we're more than happy to attack the mountain of work in the Pioneer issues backlog, given that it's officially confirmed by JSG that the workers are paid agreed hourly rate.

**Comment by bedeho** [comment]
> As the Lead of community builders squad, I don't see how the proposed scheme can be put in effect without prior settling of the working conditions and wages. With the current 6,000,000 tJOY / week budget ($120 in equivalent), we can't do much.

I just wanted to clarify one thing: how are people comparing $JOY vs $tJOY in their compensation? So lets say $1 worth of $JOY, using last valuation of token, how much are you or people in the group valuing that as $tJOY equivalent? Same? more? less? and how much either way?

**Comment by traumschule** [comment]
It seems you're asking for opinions about community developers being paid mainly via $JOY allocation instead of also $tJOY, or for a rather symbolic amount of 30 $tJOY per hour (not 1.2M according to current rate).
This would open up all current group budgets for debate because workers and leads get paid roughly what the council believes the work is worth right now (not in X years). However either way is fine for me.

**Comment by bwhm** [comment]
@dmtrjsg 

Hey,
Thanks! I have a couple of things to clarify:

> Like the idea behind splitting to two lanes. Ok we can try with 6hrs and see where we end up with!
What do you mean with 6h?

> There’s a list of 100 tasks+ so that’s a solid afternoon there on its own!
Yeah, but I don’t think we should target anything near that for the backlog (initially). I’d say we have 20-30 tops.

> So we'd need to have a separate branch for community.. I guess Theo would need to think how to handle the wf on the branches and merges lvl.

No, that's not what I mean here! The worker will work on their own fork, and make a PR to one `olympia` or something. If, say Theo, is the reviewer in this scenario, he would then make a PR to the branch in the workers fork, so if they merge it, it will be included in the PR to `olympia`!

**Comment by bedeho** [comment]
> It seems you're asking for opinions about community developers being paid mainly via $JOY allocation instead of also $tJOY, or for a rather symbolic amount of 30 $tJOY per hour (not 1.2M according to current rate).
This would open up all current group budgets for debate because workers and leads get paid roughly what the council believes the work is worth right now (not in X years). However either way is fine for me.

Everyone is primarily paid in $JOY, right now, that is incentives v3. But I think people have just not understood this yet.
It intentional that the council controls how much each group receives in $tJOY, and thus effectively $JOY, this is the whole point of the system.

**Comment by traumschule** [comment]
`I think people have just not understood this yet` - because you did not update https://www.joystream.org/token yet.

**Comment by thesan** [comment]
@bwhm thanks for this workflow looks good to me. I have just a few questions/suggestions:
 
## Community

- A minor thing to add to the step 3: the WG Lead should probably take care of checking and approving the [GH Workflow](https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks) as soon as the PR is posted.

- Should there be a round of QA before the step 5 ? The automated Vercel deployment works with PRs from public forks too, so this wouldn't require any manual deployment.

- What do you think of these slight variations on the community PR management step (board/GH workflow management excluded):
   - Step 3: First the worker open a draft PR, links it to the correct issue, and notify the Lead. Then the Lead labels it `community-dev`
   - Step 4: Once the PR is ready the worker set it ready for review. Then the internal review cycle starts
   - Step 5: Once the PR is approved by the WG Lead the PR is assign to a developer in the org team

## Jsgenesis - Review

About 3.iii. we should ask access to the Worker for access to push changes directly on there fork, is that correct ?

**Comment by dmtrjsg** [comment]
@thesan @bwhm  here's my attempt to summarise our thread visually  [Link to Miro>](https://miro.com/app/board/uXjVOAJpORw=/ )  I've added postits in red to the areas that require more syncing.  Feel free to add your thoughts as comments straight to the board, or we can talk through it on a call if it works better.
